src/Workspaces/Core/Portable/Workspace/Solution/SourceGeneratedDocumentState.cs

```diff
@@ -189,6 +189,25 @@ public SourceGeneratedDocumentState WithGenerationDateTime(DateTime generationDa
             generationDateTime);
     }
 
+    public SourceGeneratedDocumentState WithSyntaxRoot(SyntaxNode newRoot)
+    {
+        var newTreeVersion = GetNewTreeVersionForUpdatedTree(newRoot, GetNewerVersion(), PreservationMode.PreserveValue);
+        var newTreeSource = SimpleTreeAndVersionSource.Create(new TreeAndVersion(newRoot.SyntaxTree, newTreeVersion));
```

Comment:
possibly important (but probably not).  This will always create a new doc, even if the newRoot matches the old root of the DocState.  we should see here if hte current root is synchronously available (it woudl be if it was the same root), then check if it's the same, and if so, return 'this'.

This would be good to ahve a test on as well.